File: C++/Blind75/hash/128_Longest_Consecutive_Sequence.js
/**
 * Time: O(N), Space: O(N)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
    const elements = new Set(nums);

    var count = 0;
    var maxCount = 0;
    for (let n of elements) {
        if (!elements.has(n - 1)) {
            count = 1;
            var temp = n;
            while (elements.has(temp + 1)) {
                count++; temp++;
            }
            maxCount = Math.max(maxCount, count);
        }
    }

    return maxCount;
};

File: Problem-lists/Algorithm_Patterns/Hash Table/128_Longest_Consecutive_Sequence.cpp
#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <vector>
using namespace std;

/*
    - Given an unsorted array of integers nums
    - Return the length of the longest consecutive elements sequence.
    Time: O(N), Space: O(N)
*/

/*
    Note:
    - put all numbers into unordered_set -> check membership in O(1) time
    - for each number:
        - only start counting if it’s the beginning of a sequence
    - this is O(n) time because it skips all # that isn't the start => only process each number once
*/

int longestConsecutive(vector<int>& nums) {
    unordered_set<int> hash;
    for (auto n : nums) {
        hash.insert(n);
    }

    int count = 0;
    int maxCount = 0;
    for (auto n : hash) {
        if (hash.count(n - 1) == 0) {  // if is not in a consecutive sequence
            count = 1;
            int temp = n;
            while ((hash.count(temp + 1) > 0)) {
                count++;
                temp++;
            }
            maxCount = max(maxCount, count);
        }
    }

    return max(maxCount, count);
}